src/algebra/module.lean

```diff
@@ -179,61 +208,70 @@ instance semiring.to_semimodule [semiring R] : semimodule R R :=
   one_smul := one_mul,
   zero_smul := zero_mul,
   smul_zero := mul_zero }
+end
 
 @[simp] lemma smul_eq_mul [semiring R] {a a' : R} : a • a' = a * a' := rfl
 
-instance ring.to_module [ring R] : module R R :=
-{ to_semimodule := semiring.to_semimodule }
-
-/-- A ring homomorphism `f : R →+* S` defines a semimodule structure by `r • x = f r * x`. -/
-def ring_hom.to_semimodule [semiring R] [semiring S] (f : R →+* S) : semimodule R S :=
+/-- A ring homomorphism `f : R →+* M` defines a module structure by `r • x = f r * x`. -/
+def ring_hom.to_semimodule [semiring R] [ring M] (f : R →+* M) : semimodule R M :=
+semimodule.of_core
 { smul := λ r x, f r * x,
   smul_add := λ r x y, by unfold has_scalar.smul; rw [mul_add],
   add_smul := λ r s x, by unfold has_scalar.smul; rw [f.map_add, add_mul],
   mul_smul := λ r s x, by unfold has_scalar.smul; rw [f.map_mul, mul_assoc],
-  one_smul := λ x, show f 1 * x = _, by rw [f.map_one, one_mul],
-  zero_smul := λ x, show f 0 * x = 0, by rw [f.map_zero, zero_mul],
-  smul_zero := λ r, mul_zero (f r) }
-
-/-- A ring homomorphism `f : R →+* S` defines a module structure by `r • x = f r * x`. -/
-def ring_hom.to_module [ring R] [ring S] (f : R →+* S) : module R S :=
-{ to_semimodule := f.to_semimodule }
+  one_smul := λ x, show f 1 * x = _, by rw [f.map_one, one_mul] }
 
-/-- A class saying that `f` is an `R` linear map. Though it is a class,
-it is used as an explicit argument in most lemmas. -/
+/-- A map `f` between semimodules over a semiring is linear if it satisfies the two properties
+`f (x + y) = f x + f y` and `f (c • x) = c • f x`. The predicate `is_linear_map R f` asserts this
+property. A bundled version is available with `linear_map`, and should be favored over
+`is_linear_map` most of the time. -/
 class is_linear_map (R : Type u) {M : Type v} {M₂ : Type w}
```

Comment:
`git grep '\[is_linear_map' ` returns zero matches, so this can be safely changed to `structure`.

Reply:
Done, thanks

Reply:
In fact, it breaks something in category theory, so I reverted it.